src/pip/_internal/req/req_install.py

```diff
@@ -290,7 +297,7 @@ def hashes(self, trust_internet: bool = True) -> Hashes:
         """
         good_hashes = self.hash_options.copy()
         link = self.link if trust_internet else self.original_link
-        if link and link.hash:
+        if link and link.hash and link.hash_name:
             good_hashes.setdefault(link.hash_name, []).append(link.hash)
```

Comment:
Hmm, I wonder if this is actually an error in typing and `good_hashes` should really be typed as `Dict[Optional[str], List[str]]` instead.

Reply:
It doesn't make sense to have `None` be the hash algorithm.

Reply:
This should be `assert link.hash_name is not None` or equivalent here, instead of changing the if statement here.